domain/machine/state/machine_cloud_instance.go

```diff
@@ -148,3 +149,52 @@ WHERE machine_uuid=$instanceTag.machine_uuid
 		return nil
 	})
 }
+
+// InstanceId returns the provider specific instance id for this machine.
+// If the machine is not provisioned, it returns a NotProvisionedError.
+func (st *State) InstanceId(ctx context.Context, machineId string) (string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	machineIDParam := sqlair.M{"machine_id": machineId}
+	query := `
+SELECT instance_id AS &instanceID.*
+FROM machine AS m
+		JOIN machine_cloud_instance AS mci ON m.uuid = mci.machine_uuid
+WHERE m.machine_id = $M.machine_id;
+`
+	queryStmt, err := st.Prepare(query, sqlair.M{}, instanceID{})
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	var instanceId string
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		result := instanceID{}
+		err := tx.Query(ctx, queryStmt, machineIDParam).Get(&result)
+		if errors.Is(err, sqlair.ErrNoRows) {
```

Comment:
These errors should be handled outside of the transaction. We should do our best to do as little as possible in a transaction to free up resources for the rest of Juju.